### PR TITLE
Modify how libutils finds libraries

### PR DESCRIPTION
**Pull request by spxiwh**
I've been seeing some weird cases where `pip install mkl` followed by `import pycbc.fft.mkl` reports that mkl is not present in the environment. This appears to be due to how libutils is trying to find libraries, and the weirdness in how python does this, especially when conda, or venvs are in play.

The current system looks in LD_LIBRARY_PATH, pkg_config paths and an explicit path if a venv is being used. This doesn't work for conda, doesn't work if the user is doing something like spawning a venv from a conda env and doesn't work for the unittest environment (I'm not exactly sure what tox is doing!)

I tried some simple ways to patch around this, but it didn't seem to work. A more hacky solution, like the one below, does seem to work. I think we need to fix this problem because we do want `pip install mkl` to work. For e.g. a lot of users now use things like SciServer, currently the *only* workable FFT option there is numpy! The more "simple" patch covers many of the use cases, but didn't work for the test suite, so I went with this. Other options for achieving the same goal would be welcomed! Perhaps we need to explicitly ask conda/pip if it has installed the file we are looking for.

.... Also on macs the dylib files were not matching the globs being used (libmkl_rt.1.dylib didn't match), I added support for that file-naming as well.

### CODE REVIEW

File: pycbc/libutils.py
Comment (ahnitz) [inline pycbc/libutils.py:176]: @spxiwh We don't support windows atm anyway. Some of our dependencies don't support it at all right now (at least in released versions).

File: pycbc/libutils.py
Comment (ahnitz) [inline pycbc/libutils.py:179]: So this seems to pull in any path which included 'lib' in the name. I guess this makes sense.

**Review by ahnitz — APPROVED** [review]

**Comment by josh-willis** [comment]
@spxiwh So I'd like to know a little bit more about the issues you are seeing that are leading you to propose these changes. When you do `pip install mkl` I would only expect that to work if you either:
1. Source the necessary script, based on where you installed MKL, as is done in [these lines](https://github.com/gwastro/pycbc/blob/master/tools/docker_build_dist.sh#L81-L94) for our CVMFS virtual environments.  This should put the necessary path on `LD_LIBRARY_PATH`, and is how Intel seems to want users of MKL to set it up.  Have you tried this, and it isn't working?
2. Otherwise, it should be findable once you regenerate the shared library cache. This isn't done by `pip` because it's not a general purpose tool, but it is why we have [this line](https://github.com/gwastro/pycbc/blob/master/Dockerfile#L30) in our Dockerfile.

As regards conda environments, MKL is available directly in conda, apart from the use of `pip`.  Is there a reason you are not using that installation? If you've tried, how have you checked that it is not working? Again, as far as I know, you always need to activate MKL via sourcing the script above (as appropriate for whichever install of MKL you have made) but in the case of conda, @duncanmmacleod would know better than me.

**Comment by ahnitz** [comment]
> @spxiwh So I'd like to know a little bit more about the issues you are seeing that are leading you to propose these changes. When you do `pip install mkl` I would only expect that to work if you either:
> 
> 1. Source the necessary script, based on where you installed MKL, as is done in [these lines](https://github.com/gwastro/pycbc/blob/master/tools/docker_build_dist.sh#L81-L94) for our CVMFS virtual environments.  This should put the necessary path on `LD_LIBRARY_PATH`, and is how Intel seems to want users of MKL to set it up.  Have you tried this, and it isn't working?
> 2. Otherwise, it should be findable once you regenerate the shared library cache. This isn't done by `pip` because it's not a general purpose tool, but it is why we have [this line](https://github.com/gwastro/pycbc/blob/master/Dockerfile#L30) in our Dockerfile.
> 
> As regards conda environments, MKL is available directly in conda, apart from the use of `pip`. Is there a reason you are not using that installation? If you've tried, how have you checked that it is not working? Again, as far as I know, you always need to activate MKL via sourcing the script above (as appropriate for whichever install of MKL you have made) but in the case of conda, @duncanmmacleod would know better than me.

@josh-willis  This is to use the MKL python distribution. No sourcing can be required if this is to work in a lot of environments. It needs to be able to find the shared object code *without* any environment variables being set and *without* pkg-config as they don't create those files unfortunately in their pypi release.

**Comment by spxiwh** [comment]
> @spxiwh So I'd like to know a little bit more about the issues you are seeing that are leading you to propose these changes. When you do `pip install mkl` I would only expect that to work if you either:
> 
> 1. Source the necessary script, based on where you installed MKL, as is done in [these lines](https://github.com/gwastro/pycbc/blob/master/tools/docker_build_dist.sh#L81-L94) for our CVMFS virtual environments.  This should put the necessary path on `LD_LIBRARY_PATH`, and is how Intel seems to want users of MKL to set it up.  Have you tried this, and it isn't working?
> 2. Otherwise, it should be findable once you regenerate the shared library cache. This isn't done by `pip` because it's not a general purpose tool, but it is why we have [this line](https://github.com/gwastro/pycbc/blob/master/Dockerfile#L30) in our Dockerfile.
> 
> As regards conda environments, MKL is available directly in conda, apart from the use of `pip`. Is there a reason you are not using that installation? If you've tried, how have you checked that it is not working? Again, as far as I know, you always need to activate MKL via sourcing the script above (as appropriate for whichever install of MKL you have made) but in the case of conda, @duncanmmacleod would know better than me.

@josh-willis There is no source script added when using the version of mkl installed through pip *or* through conda. I'm not quite sure what the second step does, but it isn't possible to perform without root permission (which is the use-case where this becomes important). As far as I can see `pip install mkl` and `conda install mkl` add the same set of .so files, the issue in both cases is that this location is not checked when trying to identify where the mkl .so file is ... The same problem would, I think, also exist with FFTW installed via conda.

.... But (digging back a bit), I think you *already* identified and tried to fix exactly this problem with #3465 the issue was that that fix only works when using pip within a virtualenv. This patch *extends* #3465 to also work when conda is involved and for whatever weird setup the unittesting is doing.

File: pycbc/libutils.py
Comment (spxiwh) [reply pycbc/libutils.py:179]: Yeah, hacky, I know. I did try alternatives, this was the only thing that worked reliably. I'm not sure how to identify all the relevant paths that need checking without doing this .... I'd be happy with something that doesn't solve the "venv built off of a conda env looking for MKL within conda" case, but does work for the unittests. Maybe `sys.base_prefix` would do that ... Not sure I'm understanding the documentation for that.

**Comment by josh-willis** [comment]
@ahnitz I understood that you are asking about the pip installation of MKL.  However it is not correct to say "No sourcing can be required." In particular, for a virtual environment, if you haven't sourced it, you are not in it, and it would be wrong in that case to return a `ctypes.CDLL` from a virtual environment that the code happened to find.  That is the motivation for the code that searches the environment for `${VIRTUAL_ENV}/lib`.

For me, that is working just fine: if I create a new virtual environment and install `pycbc` and `mkl` into it, then the existing code works.  The following also work:
1. Without using `pycbc` or its machinery at all, but inside a conda environment:
```
source /cvmfs/oasis.opensciencegrid.org/ligo/sw/conda/etc/profile.d/conda.sh
conda activate igwn-py38
python
>>> import ctypes
>>> from ctypes.util import find_library
>>> find_library('libmkl_rt.so')
'/cvmfs/oasis.opensciencegrid.org/ligo/sw/conda/envs/igwn-py38/lib/libmkl_rt.so'
```
2. Using `pycbc` inside of a conda install:
```
source /cvmfs/oasis.opensciencegrid.org/ligo/sw/conda/etc/profile.d/conda.sh
conda activate igwn-py38
python
>>> import pycbc
>>> from pycbc.libutils import get_ctypes_library
>>> get_ctypes_library('libmkl_rt.so','mkl')
<CDLL '/cvmfs/oasis.opensciencegrid.org/ligo/sw/conda/envs/igwn-py38/lib/libmkl_rt.so', handle 56268e320820 at 0x7fe354658f40>
```
I also think that what Ian tried and said didn't work (namely, replacing the search inside of `${VIRTUAL_ENV}/lib` with one over `os.path.join(sys.prefix,'lib')` definitely *should* have worked, and would be something I'm happy to accept as a more general improvement over the code that's there now.

@spxiwh Which cases exactly didn't work for you with that fix?  I want to be careful about a couple of things:
1. Installing via `pip` inside a `conda` environment.  I think in this case what we want is to install MKL via conda, and use that.  As I indicated above, for me, this was working.
2. The test setup.  Not that I don't care that it works, but I want to be exactly sure what it is doing.  I thought this used to be done through one of the docker containers---did that change?  I want to be wary of breaking things just from a "let's get something that makes the CI test work" if we aren't understanding what about it is causing things to fail.  Likewise with some of the other platforms you reference.  I've never used SciServer, for example, and I actually *don't* agree that we just want `pip install mkl` to always work.  In some cases what we probably should be doing is installing MKL some other way (conda, package manager into Docker or Singularity) and then making PyCBC aware of that.  This is a concern both about environment fragility, and ensuring that we are following the licensed use of the software appropriate to each way of "containerizing" (broadly speaking) our code.

**Comment by ahnitz** [comment]
> @ahnitz I understood that you are asking about the pip installation of MKL. However it is not correct to say "No sourcing can be required." In particular, for a virtual environment, if you haven't sourced it, you are not in it, and it would be wrong in that case to return a `ctypes.CDLL` from a virtual environment that the code happened to find. That is the motivation for the code that searches the environment for `${VIRTUAL_ENV}/lib`.
>

One of the use cases is a hosted environment, in which case there may be no opportunity to do any sourcing, only access to conda / pip. That's what I was referring to. In those case, we don't want sourcing anything to be required as it's not really an option.

**Comment by josh-willis** [comment]
> > @ahnitz I understood that you are asking about the pip installation of MKL. However it is not correct to say "No sourcing can be required." In particular, for a virtual environment, if you haven't sourced it, you are not in it, and it would be wrong in that case to return a `ctypes.CDLL` from a virtual environment that the code happened to find. That is the motivation for the code that searches the environment for `${VIRTUAL_ENV}/lib`.
> 
> One of the use cases is a hosted environment, in which case there may be no opportunity to do any sourcing, only access to conda / pip. That's what I was referring to. In those case, we don't want sourcing anything to be required as it's not really an option.

As I wrote, if you're using conda, you should install MKL via conda.  What other hosted environment are you considering? Please be precise, explaining what software the hosting environment is using, in detail.

**Comment by spxiwh** [comment]
@josh-willis Thanks for the response. I do agree that we want to be clear about what types of environment etc. we want to support. I think that I was using a poorly setup conda env on my local machine when testing this, that lead to my statements about conda. When doing what you describe I also see the library loaded correctly.

This still is not working on the SciServer hosted environment. What I think is going on there is that conda is being used, but probably not properly initialised internally? I'm not really sure, we *don't* get to control that, we can just use pip and conda to install things there. This use case *is* fixed by using `sys.prefix` instead of checking for the VIRTUALENV environment variable, so I think we should make that change. I do emphasize though that is important to respond to what users are doing. Tools like SciServer (and Google Colab, and amazon/microsoft equivalents) are being used increasingly for tasks that don't require a huge amount of computing (and even some that do, SciServer offers a run-in-the-background service alongside the noteboooks). I had (for e.g.) a project student this Spring who's code ran much slower than it should have done, because it was FFT limited and the numpy FFT available on SciServer seems to be very slow. I would therefore advocate that we *do* take reasonable steps to support these environments.

I will try and explore further what's going on with the unit-tests. It's not always easy to figure out exactly what's going on internally with github actions though ...

**Comment by spxiwh** [comment]
@josh-willis It looks like I needed to use `sys.base_prefix` instead of `sys.prefix` for the unittests to work correctly. The hacky block is now removed and hopefully this is okay now and in-line with the discussion above.

**Comment by spxiwh** [comment]
I note that the tests here aren't really testing this yet, as there's no test code that needs this. The code in #3742 does need this though for the mac tests, so the tests there (which should now be passing) are meaningful.

**Comment by josh-willis** [comment]
@spxiwh What tests precisely have you done?  If I try in a `python2.7` venv, then `base_prefix` doesn't seem to exist. Is this only on Macs that it's a problem?

I reiterate my over-arching concern here: this feels like we are trying to make one way of installing MKL always work, even when it's the wrong way.  In almost every case we should be expecting users to be able to install MKL (or have it installed for them) in their "system," whether that be an actual system install, Docker or Singularity container, or conda environment. If they are then wanting to run a virtual environment on top of that, then that venv should be able to find and use the "system" install.  In which cases that we know of is this approach not working, and what does `sys.prefix` point to in such cases?

**Comment by spxiwh** [comment]
@josh-willis I've tested this on linux, mac and on SciServer as well as with the unittesting in #3742 .... but always with python3. I see we're also not doing any testing on mac environments in python3 any more, so the fact that `sys.base_prefix` doesn't exist in python2 was something I missed. I'm not too worried about keeping python2 support as we plan to retire that completely in the near future, but could add a short-term if switch to use `sys.prefix` on python2, *if* we agree what's right to do here in the general python3 case.

To followup on one point. I *don't* think that `pip install mkl` is a wrong thing do ... what am I missing? I've compared the pypi and conda mkl packages, they install *the same* set of files. Why should one be right, and one be wrong, especially as our documentation encourages pip over conda at the moment? ..... That said, even on SciServer doing `conda install mkl` doesn't work unless we add the change to use sys.prefix *or* sys.base_prefix (which point to the same place).

The case which motivated me to swap to `sys.base_prefix` was the case of the unittesting. This sets up linux/mac environments and then uses `tox py-unittest` to run the tests. This seems to setup another venv (or similar) on top of what's already being used. In that case using `sys.base_prefix` does point to right location, but `sys.prefix` does not. I have not found an example where `sys.base_prefix` does not point to where we want it to, but I couldn't really join up the documentation with the behaviour I observed for these variables.

We do need to figure out a way where our unittests work, on both mac and linux, with MKL, as well as solving the issue that people should be able to install and use MKL without needing system priviliges (i.e. I want `pip install mkl` and `conda install mkl` to "just work"). I'm not really sure of the difference between sys.prefix, and sys.base_prefix and which is best to use, but can only report that one works for all the use-cases I've seen, and the other doesn't. If the "right thing" to do is to use `sys.prefix` then we need a recipe to make MKL available for our mac unittests. I haven't found an alternative solution to that. Any suggestions?

**Comment by josh-willis** [comment]
@spxiwh So I've done some testing (all on Linux, in fact just a couple of different head nodes at CIT) of various permutations of using `conda`, `pip`, and `virtualenv`.  I've done all of this in Python 3.8; while I'm wary of breaking Python 2 support until we can drop it from our production uses, as you say, that's imminent, so at least in this comment I'll ignore that.

I've in some sense always started from a `conda` environment cloned from one of our IGWN distributions.  Specifically, I've done:
1.  `conda activate igwn-py38`
2. `conda env export >> igwn-py38.yml`
3.  `touch .noconda`, logout and log back in with a miniconda installation I already had (this is so that I can create a fully writeable conda environment that should be an exact duplicate of the CVFMS distributed igwn environment, at least until I start changing it)
4. `conda env create -f igwn-py38.yml -p $PWD/name_my_conda_dir_something`
5. `conda activate $PWD/name_my_conda_dir_something`

I did this twice: once to make a clean copy, and once to make a copy into which I was going to `pip install mkl`, even though the conda environment itself already had mkl installed from the conda package. From each of these two conda enviornments, I also created inside of them two different virtual environments, to test some other things as I'll describe below. For each environment I did the following test:
```
>>> import pycbc
>>> from pycbc.libutils import get_ctypes_library
>>> get_ctypes_library("mkl_rt","mkl")
<CDLL '/home/joshua.willis/projects/pycbc/tests/mkl-installs/conda_standalone/venv_pip_install_mkl/lib/libmkl_rt.so.1', handle 560e0f5e1370 at 0x7fc62d650940>
```
That shows what happens when it works; when it doesn't, the last line returns nothing, because it hasn't found the library.  With that overview, here's what I tried, and what I found:
1. Inside of the cloned conda environment, where both pycbc and mkl were installed as conda packages. This worked for me.
2. Inside of a cloned conda environment, where pycbc and mkl were installed as conda packages, but I then went and installed mkl again via pip, into the conda environment. This worked for me.
3. Inside of number (1) above, I created a new virtual environment (using the `virtualenv` from the conda environment). I activated this virtual environment and then populated it only with what `pip install pycbc mkl` put into it.  This worked for me.
4. Inside of number (2) above, again create a new virtual environment, populate it with `pip install pycbc mkl`. This worked for me.
5. Inside of number (1) again, yet another virtual environment, this time populated only with `pip install pycbc`.  This did *not* work.
6. Inside of number (2) again, again a brand new virtual environment, and again populated only with `pip install pycbc`. Just as with scenario (5), this did *not* work.

When I did a hasty web search about `sys.base_prefix`, it lead me to [this](https://docs.python.org/3/library/sys.html#sys.base_prefix):
```
 sys.base_prefix

    Set during Python startup, before site.py is run, to the same value as prefix. If not running in a virtual environment, the values will stay the same; if site.py finds that a virtual environment is in use, the values of prefix and exec_prefix will be changed to point to the virtual environment, whereas base_prefix and base_exec_prefix will remain pointing to the base Python installation (the one which the virtual environment was created from).

    New in version 3.3
```
When I examined both `sys.prefix` and `sys.base_prefix` inside the six scenarios I tested above, I think what I saw was always consistent with this documentation: the two pointed to the same thing if I was only in a conda environment, but inside of a virtual environment, `sys.prefix` pointed to the virtual environment and `sys.base_prefix` pointed to the conda environment from which that virtual environment was created.

Given that, I really *don't* want the change that looks at `sys.base_prefix`; it should look at `sys.prefix`. The latter is pointing at whichever Python environment I'm trying to use; the former at whatever environment I created that from.  And as a user who has chosen to create a virtual environment, it would be clearly incorrect behavior if our package were to look outside that environment to pull in a library that I hadn't installed into my virtual environment; it would be breaking the isolation that a virtual environment is supposed to create. In fact, if we were to modify this as your current patch does, then if I had installed a library into my virtual environment, but there was a different version of it in the "base" (which would be conda in my examples above, but could in general be a system install if I just created a virtual environment from the command line on a linux system), I would get that base version and not the version I had purposefully installed into my virtual environment.

So, even if it doesn't solve all of our current problems, for the reasons above, I can't accept a patch that looks at `sys.base_prefix` instead of `sys.prefix`.  Even if it were to look first at the latter and fall back to the former, I think it would still be wrong: if I haven't installed my package in my virtual environment, it *should* fail if I try to use it.

What I can't tell is how consistent what I outline above is with some of the other scenarios you are describing.  But if it's not, we need to find out why not, and figure out the right way to address that. In many cases I really can't tell what exactly has been done: it's not enough to say `pip install mkl` didn't work---I need to know if that was into a conda environment and conda was all I was using (in which case, I would *not* expect it to fail) or I did it into a conda environment, then created inside of that a virtual environment, but that venv didn't have mkl (in which case I *would* expect it to fail).  So, some things that maybe are happening (but maybe aren't):
1. What is SciServer doing? What API does it provide and how are you testing it? Does it sound like it's consistent with the cases I outlined above? If not, maybe they are doing something behind the scenes? (for example: "Oh, you tried to pip install mkl, but mkl is already there, so I'll just ignore it")
2. I don't know what role, if any, OSes other than Linux are playing here.
3. I can see from the documentation that `tox` runs tests in a virtual environment, but nothing about how it constructs that environment. Is it looking at `requirements.txt`? How does it handle optional dependencies---do we need to be telling it to install mkl because it's purely optional for PyCBC but we want it when we're trying to test MKL? In what kind of environment are we starting tox? Another `virtualenv`, so that we have a `virtualenv` in a `virtualenv`? Or is it a `virtualenv` in a `conda` environment, or a Docker image, or . . .

I won't have much more time to look at this over the next few days; I've put a lot of urgent things to the side to do the work I've already done.  But this MR isn't there yet.

**Comment by spxiwh** [comment]
Hi Josh,

Thanks for the detailed reply. There's a lot of stuff in this reply, but the short version of this is that I agree that using *only* sys.base_prefix was a bad idea. I'm not so convinced that the model of checking sys.prefix *and then* checking sys.base_prefix is a bad idea. However, I'm not going to make that case today, and the patch is now changed such that instead of using the VIRTUAL_ENV variable, we search sys.prefix instead. I think I now have a better understanding of what tox is doing for the unittests and will resolve this problem in a different way.

To go into some more detail in the differences between SciServer and CIT. Under the hood SciServer runs a SL7 Linux system with an environment managed by conda. I don't think that conda is setup "properly" as it doesn't run `conda activate` internally, but instead adds the relevant path to the PATH variable so that the python executable that is used is the appropriate one. Nevertheless, nonexpert users will not be able to change this. While SciServer is primarily a Jupyter notebook server it does allow the user to open a terminal, and the below is based upon opening a terminal and then running conda activate "properly" to avoid the possibility that this was the difference. After doing this I still cannot find a locally installed mkl (installed with conda install mkl) *before* this patch. 

I have identified what the problem, and some differences, are. These are all in the ctypes.util module. The way this locates modules is a bit convoluted but it basically tries 4 things:

* Look in sys.prefix/lib
* Use `gcc -Wl,-t -o /tmp/RANDOMNAME -lmkl_rt` and parse that to get the library path
* Use `ld -t -o /dev/null -lmkl_rt` and parse that to get the library path
* Use `/sbin/ldconfig -p` and parse that to get the library path.

For a "standard" conda setup the last 3 will not find the MKL library (they *might* if gcc is installed within the conda environment). These will only search system default paths, and anything in LD_LIBRARY_PATH (which conda and venv do *not* set ... I believe they instead add the necessary paths as -L flags during compilation if needed but am not 100% sure on that). So then we're back to using sys.prefix. This would be great, if it worked, but it doesn't work reliably. The version of ctypes.util on SciServer does something bizarre here. It first finds the library (in sys.prefix), then runs `objdump -p -j .dynamic libmkl_rt.so` on the file, it greps the output for the string "SONAME", which gives us the name of the so file (which we already know), and then it searches for it again. On SciServer the output of objdump command *does not* contain a SONAME entry. On LHO/CIT it *does* contain a SONAME entry. I don't know why this is different. However, the version of ctypes.util on LHO/CIT also doesn't include this complexity and just finds the library, so that version *would* work on SciServer. In summary I think this code seems a bit unreliable when looking under sys.prefix. (Although if this code can work reliably, it could remove the need to do the "get_libpath_from_dirlist" check entirely, as it does seem that this block of code, and the ctypes function are largely trying to do the same thing. ctypes supports more OS versions, but doesn't seem to work reliably ... I think that's a bigger patch though. I'm now considering this just a "Extend the `if "VIRTUAL_ENV in os.environ` check to be more generic" patch)

The second issue was around tox and the unittests. I've understood now that tox does *not* use the venv that we install to do the testing (as I had thought it did), but instead builds *yet another* venv. There might be scope to reduce duplication in the testing because of this, but I won't look to address that here. I reproduced the issue on a linux machine, and figured out how to get mkl installed within the tox venv. I'll include that in #3742, not here.

Thanks again

**Review by josh-willis — APPROVED** [review]
@spxiwh I'll say a bit more in Slack, but a couple of quick things here.  But as written I think this patch is now OK, and we can look at the unittest issue elsewhere.
1. I do think we have an obligation to not look in `sys.base_prefix`, because that will only be different from `sys.prefix` inside of a virtual environment.  Virtual environments are supposed to be isolated.  Normally we don't have to do anything special to ensure that, but because we are explicitly poking around trying to find things in this module, we do have to be careful.  To pull in a library that was in `sys.base_prefix` but not `sys.prefix` could be rightly regarded as a bug by users of our software, and I want to avoid that.
2. I agree that what `ctypes` does to find libraries is necromancy and fragile.  I suspect it's maintainers might agree. As a note, one of the main reasons for this module originally was the desire to be able to find things that are specified via `pkg-config`.  As far as I know none of `conda`, `virtualenv/venv`, nor `ctypes` care anything about that, even though it's a standard Linux tool specifically designed for "how do I find where my libraries and headers are installed, based on the preference order of the user."

**Comment by spxiwh** [comment]
Thanks @josh-willis! Merging this now. The unittests are now passing with this change applied in #3742, so I think that's all good too. (Just need to rebase 3742 after merging this).

**Pull request merged**
